_posts/2023-11-20-unlocking-the-secrets-to-ingestion.md

```diff
@@ -0,0 +1,191 @@
+
+We recommend setting the translog flush size to around 25% of your available Java heap defined earlier. On a node with 16GB of memory we would set our heap to 50% of the available memory which would be 8GB for the heap. Then we would take 25% of the heap to set aside for the translog flushes or 2GB.
+
+```json
+POST index-name/_settings
+{
+	"index" {
+		"translog.flush_threshold_size": "2048MB"
+	}
+}
+```
+
+Small note, the translog is not in memory. We set the translog threshold  to 25% of our memory because when the translog flushes it will transfer the data from disk into memory to be compiled into a Lucene segment to be persisted to disk.
+
+# Sharding and Replication
+
+We talked a little bit about threading earlier and now we will dive a little bit more into mechanisms for making things happen in parallel. OpenSearch uses shards to partition indexes. Which shard an index goes to is based off a hash of it’s `_id`. There are primary shards and replica shards. Primary shards are responsible for ingestion and sending the data to the replicas.
+
+### Primary Shard Distribution
+
+We want to ensure we prioritize balancing our shards across nodes so the workload is evenly distributed. Without setting this we could end up with a situation like the one below where all our primary shards for end up on the same node. That would cause our Node 1 here to be completely saturated while the other nodes may be underutilized.
+
+![Primary Shard Distribution](assets/media/blog-images/2023-11-20-unlocking-the-secrets-to-ingestion/PrimaryShardDistribution.svg){: .img-fluid }
+
+```json
+PUT _cluster/settings
+{
+	"persistent": {
+		"cluster.rounting.allocation.balanace.prefer_primary": True
+	}
+}
+```
+
+Enabling primary shard distribution ensures primary shards are given the first preference when it comes to spreading them across nodes. When this is enabled replicas may be more likely to end up on the same node however this is okay in an ingestion heavy workload.
+
+![Primary Shard Distribution Enabled](assets/media/blog-images/2023-11-20-unlocking-the-secrets-to-ingestion/PrimaryShardDistributionEnabled.svg){: .img-fluid }
+
+### Segment Replication
+
+One setting that dramatically impact ingestion is segment replication. Out of the box OpenSearch uses document replication. In this strategy data is first sent to the primary shard which then ingests it back into a Lucene segment. After it’s done that it will send the original unprocessed document to any replicas. They will then ingest the document back to their own Lucene segments. This works really well if you need more immediate consistency however it is duplicating the work done on the CPU for each replica.
```

Comment:
```suggestion
One setting that dramatically impacts ingestion is segment replication. Out-of-the-box OpenSearch uses document replication. In this strategy data is first sent to the primary shard that then ingests it back into a Lucene segment. After it’s done that it will send the original unprocessed document to any replicas. They will then ingest the document back to their own Lucene segments. This works really well if you need more immediate consistency; however, it's duplicating the work done on the CPU for each replica.
```